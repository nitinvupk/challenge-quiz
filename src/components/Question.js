import React, { Fragment, useState } from 'react'

function Question (props) {
  const [ answer, setAnswer ] = useState(null)

  const handleChooseAnswer = (index) => {
    if (answer) return
    setAnswer(props['options'][index])
  }

  const handleNextQuestion = () => {
    let isCorrect = props.correct_answer === answer
    if (isCorrect) props.handleScore()
    props.nextQuestion()
    setAnswer(null)
  }

  const displayResult = () => {
    let isCorrect = props.correct_answer === answer
    return (
      answer &&
      <div className='text-center'>
        <h1>{isCorrect ? 'Correct!' : 'Sorry!'}</h1>
        <button className='btn' onClick={handleNextQuestion}>Next Question</button>
      </div>
    )
}
  return (
    <Fragment>
      <p>{decodeURIComponent(props.question)}</p>
      {props.options &&
        <Fragment>
          <div className='buttonBox'>
            <button className={'btn ' + (answer && answer === props['options'][0] ? 'selected' : '')} onClick={() => handleChooseAnswer(0)}>
              {decodeURIComponent(props.options[0])}
            </button>
            <button className={'btn ' + (answer && answer === props['options'][1] ? 'selected' : '')} onClick={() => handleChooseAnswer(1)}>
              {decodeURIComponent(props.options[1])}
            </button>
          </div>
          {props['options'].length > 2 &&
            <div className='buttonBox'>
              <button className={'btn ' + (answer && answer === props['options'][2] ? 'selected' : '')} onClick={() => handleChooseAnswer(2)}>
                {decodeURIComponent(props.options[2])}
              </button>
              <button className={'btn ' + (answer && answer === props['options'][3] ? 'selected' : '')} onClick={() => handleChooseAnswer(3)}>
                {decodeURIComponent(props.options[3])}
              </button>
            </div>
          }
        </Fragment>
      }
      { displayResult() }
    </Fragment>
  )
}
export default Question
