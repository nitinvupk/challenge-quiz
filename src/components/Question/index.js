import React, { Fragment, useState } from 'react'

function Question (props) {
  const [ answer, setAnswer ] = useState(null)

  const handleChooseAnswer = (index) => {
    if (answer) return
    setAnswer(props['options'][index])
  }

  const handleNextQuestion = () => {
    let isCorrect = props.correct_answer === answer
    if (isCorrect) props.handleScore()
    props.nextQuestion()
    setAnswer(null)
  }

  const displayResult = () => {
    let isCorrect = props.correct_answer === answer
    return (
      answer &&
      <div className='text-center'>
        <h1>{isCorrect ? 'Correct!' : 'Sorry!'}</h1>
        <button id="nextQuestion" className='btn' onClick={handleNextQuestion}>Next Question</button>
      </div>
    )
  }
  return (
    <Fragment>
      <p>{decodeURIComponent(props.question)}</p>
      {props.options &&
        <Fragment>
          <div className='buttonBox'>
            {props['options'].map((opt, idx) => (
              <div className='btnContainer' key={idx}>
                <button className={'btn ' + (answer && answer === opt ? 'selected' : '')} onClick={() => handleChooseAnswer(idx)}>
                  {decodeURIComponent(opt)}
                </button>
              </div>
            ))}
          </div>
        </Fragment>
      }
      { displayResult() }
    </Fragment>
  )
}
export default Question
