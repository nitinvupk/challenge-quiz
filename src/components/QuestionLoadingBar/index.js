import React from 'react'

function QuestionLoadingBar (props) {
  let width = 0
  if (props.questionNumber && props.totalCount) {
    width = props.questionNumber / props.totalCount * 100
  }

  return (
    <div className='loadingBar'>
      <div className='progress' style={{ width: `${width}%` }} />
    </div>
  )
}
export default QuestionLoadingBar
