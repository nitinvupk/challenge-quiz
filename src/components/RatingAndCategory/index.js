import React, { Fragment } from 'react'

const ratingLevels = {
  'hard': { select: 3, nonSelect: 2 },
  'medium': { select: 2, nonSelect: 3 },
  'easy': { select: 1, nonSelect: 4 }
}

function RatingAndCategory (props) {
  const createRating = (difficulty) => {
    if (difficulty) {
      let ratings = []
      let selectedRating = ratingLevels[difficulty]
      for (let i = 0; i < 5; i++) {
        if (i < selectedRating.select) {
          ratings.push(<i className='fa fa-star' key={i} />)
        } else {
          ratings.push(<i className='fa fa-star ratingDisable' key={i} />)
        }
      }
      return ratings
    }
  }

  return (
    <Fragment>
      <h1>Question {props.currentQueNumber + 1} of {props.totalCount}</h1>
      <p>{decodeURIComponent(props.category) }</p>
      {
        createRating(props.difficulty)
      }
    </Fragment>
  )
}
export default RatingAndCategory
