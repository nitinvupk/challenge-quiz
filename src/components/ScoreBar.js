import React, { useState, useEffect, Fragment } from 'react'
import { calculatePercent } from '../helpers/index'

function ScoreBar (props) {
  const [ currentScorePercent, setCurrentScorePercent ] = useState(0)
  const [ allRemWrongPercent, setAllRemWrongPercent ] = useState(0)
  const [ allRemCorrectPercent, setAllRemCorrectPercent ] = useState(0)
  const [ expect, setExpect ] = useState(0)
  useEffect(() => {
    let correct = calculatePercent(props.score, props.totalCount)
    let allRemWrong = calculatePercent(props.score, props.currentQueNumber)
    let allRemCorrect = calculatePercent((props.totalCount - props.currentQueNumber + props.score), props.totalCount)

    setCurrentScorePercent(correct)
    setAllRemCorrectPercent(allRemCorrect - allRemWrong)
    setExpect(allRemCorrect)
    setAllRemWrongPercent(allRemWrong - correct)
  }, [
    props.score,
    props.totalCount,
    props.currentQueNumber
  ])

  return (
    <Fragment>
      <div className='percentBar'>
        <div className='currentScorePercent'>Score:{ currentScorePercent }%</div>
        <div className='maxScorePercent'>Max:{ expect }%</div>
      </div>
      <div className='scoreBox'>
        <div className='allWrongBar' style={{ width: `${allRemWrongPercent}%` }} />
        <div className='currentScoreBar' style={{ width: `${currentScorePercent}%` }} />
        <div className='allCorrectBar' style={{ width: `${allRemCorrectPercent}%` }} />
      </div>
    </Fragment>
  )
}
export default ScoreBar
