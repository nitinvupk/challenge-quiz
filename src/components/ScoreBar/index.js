import React, { useState, useEffect, Fragment } from 'react'
import { calculatePercent } from '../../helpers/index'

function ScoreBar (props) {
  const [ currentScorePercent, setCurrentScorePercent ] = useState(0)
  const [ lowestPossiblePercent, setLowestPossiblePercent ] = useState(0)
  const [ maximumPercent, setMaximumPercent ] = useState(0)
  const [ maxScore, setMaxScore ] = useState(0)
  useEffect(() => {
    let correct = calculatePercent(props.score, props.totalCount)
    let lowestPossible = calculatePercent(props.score, props.currentQueNumber)
    let maximum = calculatePercent((props.totalCount - props.currentQueNumber + props.score), props.totalCount)

    setCurrentScorePercent(correct)
    setMaximumPercent(maximum - lowestPossible)
    setMaxScore(maximum)
    setLowestPossiblePercent(lowestPossible - correct)
  }, [
    props.score,
    props.totalCount,
    props.currentQueNumber
  ])

  return (
    <Fragment>
      <div className='percentBar'>
        <div className='currentScorePercent'>Score:{ currentScorePercent }%</div>
        <div className='maxScorePercent'>Max:{ maxScore }%</div>
      </div>
      <div className='scoreBox'>
        <div className='lowestPossibleBar' style={{ width: `${lowestPossiblePercent}%` }} />
        <div className='currentScoreBar' style={{ width: `${currentScorePercent}%` }} />
        <div className='maximumPercentBar' style={{ width: `${maximumPercent}%` }} />
      </div>
    </Fragment>
  )
}
export default ScoreBar
