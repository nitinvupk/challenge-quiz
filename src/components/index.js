import RatingAndCategory from './RatingAndCategory'
import Question from './Question'
import ScoreBar from './ScoreBar'
import QuestionLoadingBar from './QuestionLoadingBar'

export {
  RatingAndCategory,
  Question,
  ScoreBar,
  QuestionLoadingBar
}
