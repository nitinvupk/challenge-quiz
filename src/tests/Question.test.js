import React from 'react'
import { shallow } from 'enzyme'
import { Question } from '../components'
import allQuestions from '../assets/questions.json'

let question = allQuestions[0]
question['options'] = [...question.incorrect_answers, question.correct_answer]
describe('App', () => {
  it('should render Question component', () => {
    const component = shallow(<Question {...question} />)
    expect(component).toMatchSnapshot()
  })

  it('should render title in p tag', () => {
    const component = shallow(<Question {...question} />)
    expect(component.find('p').text()).toEqual(decodeURIComponent(question.question))
  })

  it('should click button and answer should be wrong', () => {
    const component = shallow(<Question {...question} />)
    component
      .find('button')
      .at(0)
      .simulate('click')
    expect(component.find('h1').text()).toEqual('Sorry!')
  })

  it('should click button and answer should be corect', () => {
    const component = shallow(<Question {...question} />)
    component
      .find('button')
      .at(3)
      .simulate('click')
    expect(component.find('h1').text()).toEqual('Correct!')
  })
})
