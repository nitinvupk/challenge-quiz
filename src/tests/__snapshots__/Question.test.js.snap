// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render Question component 1`] = `
<Fragment>
  <p>
    What was the name of the hero in the 80s animated video game 'Dragon's Lair'?
  </p>
  <div
    className="buttonBox"
  >
    <div
      className="btnContainer"
      key="0"
    >
      <button
        className="btn "
        onClick={[Function]}
      >
        Arthur
      </button>
    </div>
    <div
      className="btnContainer"
      key="1"
    >
      <button
        className="btn "
        onClick={[Function]}
      >
        Sir Toby Belch
      </button>
    </div>
    <div
      className="btnContainer"
      key="2"
    >
      <button
        className="btn "
        onClick={[Function]}
      >
        Guy of Gisbourne
      </button>
    </div>
    <div
      className="btnContainer"
      key="3"
    >
      <button
        className="btn "
        onClick={[Function]}
      >
        Dirk the Daring
      </button>
    </div>
  </div>
</Fragment>
`;
